src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/perform_interaction_request.cc

```diff
@@ -1063,13 +1092,70 @@ void PerformInteractionRequest::SendBothModeResponse(
       msg_param.empty() ? NULL : &msg_param;
   std::string info = app_mngr::commands::MergeInfos(
       ui_perform_info, ui_info_, vr_perform_info, vr_info_);
+
   DisablePerformInteraction();
+
   SendResponse(result,
                perform_interaction_result_code,
                info.empty() ? NULL : info.c_str(),
                response_params);
 }
 
+mobile_apis::Result::eType
+PerformInteractionRequest::PrepareResultCodeForResponse(
+    const app_mngr::commands::ResponseInfo& ui_response,
+    const app_mngr::commands::ResponseInfo& vr_response) {
+  if (interaction_mode_ == mobile_apis::InteractionMode::VR_ONLY) {
+    if (first_responser_ == FirstPerformInteractionResponser::VR) {
+      return MessageHelper::HMIToMobileResult(vr_result_code_);
+    }
+  }
+
+  if (interaction_mode_ == mobile_apis::InteractionMode::BOTH) {
+    if (IsVRPerformInteractionResponseSuccessfulInBothMode()) {
+      return MessageHelper::HMIToMobileResult(vr_result_code_);
+    } else {
+      return MessageHelper::HMIToMobileResult(ui_result_code_);
+    }
+  }
+
+  return CommandRequestImpl::PrepareResultCodeForResponse(ui_response,
+                                                          vr_response);
+}
+
+bool PerformInteractionRequest::PrepareResultForMobileResponse(
+    app_mngr::commands::ResponseInfo& ui_response,
+    app_mngr::commands::ResponseInfo& vr_response) const {
+  if (interaction_mode_ == mobile_apis::InteractionMode::VR_ONLY) {
+    if (first_responser_ == FirstPerformInteractionResponser::VR) {
+      return vr_response.is_ok;
+    }
+  }
+
+  if (interaction_mode_ == mobile_apis::InteractionMode::BOTH) {
+    return (vr_response.is_ok || ui_response.is_ok);
+  }
+
+  return CommandRequestImpl::PrepareResultForMobileResponse(ui_response,
+                                                            vr_response);
+}
+
+bool PerformInteractionRequest::
+    IsVRPerformInteractionResponseSuccessfulInBothMode() {
+  using namespace mobile_apis;
+  app_mngr::commands::ResponseInfo vr_perform_info(
+      vr_result_code_, HmiInterfaces::HMI_INTERFACE_VR, application_manager_);
+  return (vr_perform_info.is_ok && interaction_mode_ == InteractionMode::BOTH);
+}
+
+void PerformInteractionRequest::StoreFirstPerformInteractionResponser(
+    FirstPerformInteractionResponser responser) {
+  first_responser_ =
+      (first_responser_ == FirstPerformInteractionResponser::NONE)
+          ? responser
+          : first_responser_;
+}
```

Comment:
@YarikMamykin Why do you make it so complicated? 
Leave it like: 
```cpp
void PerformInteractionRequest::StoreFirstPerformInteractionResponser(
    FirstPerformInteractionResponser responser) {
  if (FirstPerformInteractionResponser::NONE == first_responser_ ) {
  	first_responser_ = responser;
  }
}
```

Reply:
@AByzhynar fixed in https://github.com/smartdevicelink/sdl_core/pull/3122/commits/0190836450ab1d7e130bbb8e32315eb3325a9a68